### PR TITLE
[13/N] Use std::optional 

### PR DESCRIPTION
**Pull request by cyyever**
Follows #132361

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

File: c10/util/Optional.h [hunk 1]
```diff
@@ -10,8 +10,6 @@
 
 namespace c10 {
 // NOLINTNEXTLINE(misc-unused-using-decls)
-using std::bad_optional_access;
```
Comment (Skylion007) [inline c10/util/Optional.h:13]: Why remove the aliasees here, it could break out of tree extensions for no reason.

Comment (cyyever) [reply c10/util/Optional.h:13]: We have a series of patches to move to std::optional and we try not to break existing code as possible...

Comment (cyyever) [reply c10/util/Optional.h]: Reverted

**Review by ezyang — APPROVED** [review]

**Pull request closed without merging**